File: Algorithms/Searching/Binary Search/binary-search.go
/*
 * Binary Search
 *
 */

package main

import (
	"fmt"
)

func BinarySearch(arr []int, l, r, x int) int {
	if r >= l {
		mid := l + (r-l)/2
		if arr[mid] == x {
			return mid
		} else if arr[mid] > x {
			return BinarySearch(arr, l, mid-1, x)
		} else {
			return BinarySearch(arr, mid+1, r, x)
		}
	}
	return -1
}

func main() {
	arr := []int{3, 5, 12, 56, 92, 123, 156, 190, 201, 222}
	n := len(arr)
	x := 12
	l := 0
	r := n - 1
	fmt.Println(BinarySearch(arr, l, r, x))
}

File: Algorithms/Searching/Binary Search/binary-search.hs
{-
-- Binary Search
-}

binarySearch :: [Int] -> Int -> Int -> Int -> Int

binarySearch xs val low high
    | high < low = -1
    | xs!!mid < val = binarySearch xs val (mid+1) high
    | xs!!mid > val = binarySearch xs val low (mid-1)
    | otherwise = mid
    where
        mid = (low + high) `div` 2

main = do
    let arr = [3,5,12,56,92,123,156,190,201,222]
    print $ binarySearch arr 12 0 ((length arr) - 1)

